src/prefect/orion/orchestration/core_policy.py

```diff
@@ -450,3 +524,69 @@ async def before_transition(
             await self.abort_transition(
                 reason=f"This run cannot transition to the {proposed_state_type} state from the {initial_state_type} state."
             )
+
+
+class PermitRerunningFailedTaskRuns(BaseOrchestrationRule):
+    """
+    Permits rerunning a task run in a terminal state; enables retries and restarts.
+
+    While rerunning a flow, the client will attempt to re-orchestrate tasks that may
+    have previously failed. This rule will permit transitions back into a running state
+    if the parent flow run is either currently restarting or retrying. The task run's
+    run count will also be reset so task-level retries can still fire and tracking
+    metadata is updated.
+    """
+
+    FROM_STATES = [
+        states.StateType.FAILED,
+        states.StateType.CRASHED,
+        states.StateType.CANCELLED,
+    ]
+    TO_STATES = [states.StateType.RUNNING]
+
+    async def before_transition(
+        self,
+        initial_state: Optional[states.State],
+        proposed_state: Optional[states.State],
+        context: TaskOrchestrationContext,
+    ):
+        self.original_run_count = context.run.run_count
+        self.original_retry_attempt = context.run.flow_retry_attempt
+        self.original_restart_attempt = context.run.flow_restart_attempt
+
+        self.flow_run = await context.flow_run()
+
+        restarting = (
+            self.flow_run.run_count == 1
+            and self.flow_run.restarts > 0
+            and (self.flow_run.restarts > context.run.flow_restart_attempt)
+        )
+
+        retrying = not restarting and (
+            self.flow_run.run_count <= (self.flow_run.empirical_policy.retries + 1)
+        )
+
+        if restarting:
```

Comment:
The distinction between restarting and retrying feels arbitrary to me and also hard to internalize - is there a technical reason we need it?

Reply:
Even if we remove any bookkeeping around retrying and restarting there still is a distinction: flow run run counts are reset on a restart, otherwise they will not retry as defined, they would be the same if restarts consumed a retry, but then we'd need special handling for restarts that exceed the retry limit

Reply:
I see, so because clients can set _automatic_ retry configuration on a flow, we need a way to allow "restart from failure" that ignores that automatic retry setting

Reply:
yeah exactly

Reply:
This is still really confusing to me (and I'm not sure, even after reading the code, which is a "retry" and which is a "restart"). 

To me a "restart" is a state transition from TERMINAL to SCHEDULED that can be orchestrated at any time. In addition, the system will automatically restart flows that have been configured to do so.

Consider a flow that has no retries configured (e.g. `retries=0`). That means the system won't automatically retry the flow on failure but of course I can _manually_ put that flow back into a scheduled state to retry it if it fails. In addition, a flow with `retries=3` would automatically retry up until the third attempt, but I could still continue to manually retry it above that.

Reply:
In this model, retries are transitions from running -> scheduled, the terminal transitions were never hit for automatic retries. Restarts were intended to encompass the terminal -> scheduled case; with different behavior. As mentioned in another comment we've decided to unify running -> scheduled and terminal -> scheduled for simplicity